sway-ir/src/optimize/mem2reg.rs

```diff
@@ -148,19 +151,63 @@ pub fn compute_livein(
     result
 }
 
-/// Promote local values that are accessed via load/store to SSA registers.
-/// We promote only locals of non-copy type, whose every use is in a `get_local`
+// Promote loads of globals constants to SSA registers
+fn promote_globals(context: &mut Context, function: &Function) -> Result<bool, IrError> {
+    let mut replacements = FxHashMap::<Value, Constant>::default();
+    for (_, inst) in function.instruction_iter(context) {
+        if let ValueDatum::Instruction(Instruction {
+            op: InstOp::Load(ptr),
+            ..
+        }) = context.values[inst.0].value
+        {
+            if let ValueDatum::Instruction(Instruction {
+                op: InstOp::GetGlobal(global_var),
+                ..
+            }) = context.values[ptr.0].value
+            {
+                if !global_var.is_mutable(context)
+                    && is_promotable_type(context, global_var.get_inner_type(context))
+                {
+                    let constant = *global_var
+                        .get_initializer(context)
+                        .expect("We're dealing with an uninitialized value");
+                    replacements.insert(inst, constant);
+                }
+            }
+        }
+    }
+
+    if replacements.is_empty() {
+        return Ok(false);
+    }
+
+    let replacements = replacements
+        .into_iter()
+        .map(|(k, v)| (k, Value::new_constant(context, v)))
+        .collect::<FxHashMap<_, _>>();
+
+    function.replace_values(context, &replacements, None);
+
+    Ok(true)
+}
+
+/// Promote memory values that are accessed via load/store to SSA registers.
+/// Globals: We promote only non-mutable globals of non-copy types
+/// Locals: We promote only locals of non-copy type, whose every use is in a `get_local`
 /// without offsets, and the result of such a `get_local` is used only in a load
 /// or a store.
 pub fn promote_to_registers(
     context: &mut Context,
     analyses: &AnalysisResults,
     function: Function,
 ) -> Result<bool, IrError> {
+    let promoted_globals = promote_globals(context, &function)?;
+
     let safe_locals = filter_usable_locals(context, &function);
```

Comment:
Similar to `mem_cpy_opt`, this will be more readable if we extract promoting of locals to its own function:
```Rust
let mut modified = false;
modified |= promote_globals(context, &function)?;
modified |= promote_locals(context, &function)?;

Ok(modified)
```